### PR TITLE
Refactor part of mom_main.c

### PR DESCRIPTION
**Pull request by SudeshnaMoulik**
#### Describe Bug or Feature
Refactor some part of mom_main.c and remove few unsupported platform code.

#### Describe Your Change

- Allow floating value in waitime, which will help us to tune further mom daemon performance
- Refactor some part of mom_main.c which are platform specific and removed unsupported platform code.

#### Attach Test and Valgrind Logs/Output

3690 | sudemo | regression | Tests from given sources on platforms UBUNTU1804, SUSE15, CENTOS8, CENTOS7 | SudeshnaMoulik/openpbs@mom_perf | 14th Sep 2020 02:52:39.967 PM +05:30 | 00:49:45.528 | 1837 | 0 | 0 | 4 | 0 | 1 | 1 | 1831
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --

3694 | sudemo | regression | Rerun Only Failed Tests From #3690 | SudeshnaMoulik/openpbs@mom_perf | 14th Sep 2020 06:33:29.783 PM +05:30 | 00:05:40.851 | 5 | 0 | 0 | 4 | 0 | 0 | 0 | 1
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --

Note: those 4 test case also failed for master branch (job id: 3710)

### CODE REVIEW

File: src/resmom/mom_main.c [hunk 1]
```diff
@@ -7121,49 +6435,17 @@ static char	configusage[] = "%s -s insert scriptname inputfile\n"
...
 usage(char *prog)
 {
-#ifdef	WIN32
-	fprintf(stderr,
-		"Usage: %s [-C chkdirectory][-d dir][-c configfile][-r|-p][-R port][-M port][-S port][-L log][-a alarm][-n nice]\n", prog);
-#else
-	fprintf(stderr,
-		"Usage: %s [-C chkdirectory][-d dir][-c configfile][-r|-p][-R port][-M port][-L log][-a alarm][-n nice]\n", prog);
-#endif
+	snprintf(log_buffer, sizeof(log_buffer), "%s",
```
Comment (subhasisb) [inline src/resmom/mom_main.c:6438]: I think you should just a complete different implementation of usage for win32 and linux

File: src/resmom/mom_main.c [hunk 2]
```diff
@@ -7121,49 +6435,17 @@ static char	configusage[] = "%s -s insert scriptname inputfile\n"
...
+		"Usage: %s [-C chkdirectory][-d dir][-c configfile][-r|-p][-R port][-M port][-L log][-a alarm][-n nice]");
+	strcat(log_buffer, PLATFORM_SPECIFIC_USAGE);
```
Comment (subhasisb) [inline src/resmom/mom_main.c:6440]: This does not make sense

File: src/include/portability.h [hunk 3]
```diff
@@ -52,4 +52,19 @@
 #define dlerror_reset() dlerror()
 #define SHAREDLIB_EXT "so"
 
+#ifdef PBS_MOM
+#define TRAILING_CHAR '/'
+#define verify_dir(dir_val, isdir, sticky, disallow, fullpath) tmp_file_sec(dir_val, isdir, sticky, disallow, fullpath)
+#define STAT_CHKPT_FILE(file_name, buf) stat(file_name, buf)
+#define FULLPATH 1
+#define PLATFORM_SPECIFIC_USAGE "[-S port]\n"
```
Comment (subhasisb) [inline src/include/portability.h:60]: let's not have this, instead make two functions for usage() one for linux and other for windows

File: src/resmom/mom_main.c [hunk 4]
```diff
@@ -3949,26 +3486,15 @@ set_checkpoint_path(char *value)
...
-#endif	/* WIN32 */
+	perm = get_permission("write");
+	rc = chk_file_sec(newpath, 1, 0, perm, 0);
 #else
 	{
 		struct stat	sb;
-#ifdef	WIN32
-		/* use windowed lstat for it works */
-		if (lstat(newpath, &sb) == -1)
+		if (STAT_CHKPT_FILE(newpath, &sb) == -1)
```
Comment (subhasisb) [inline src/resmom/mom_main.c:3494]: Don't do this, instead implement lstat() in libwin

File: src/include/portability.h [hunk 5]
```diff
@@ -52,4 +52,19 @@
 #define dlerror_reset() dlerror()
 #define SHAREDLIB_EXT "so"
 
+#ifdef PBS_MOM
+#define TRAILING_CHAR '/'
+#define verify_dir(dir_val, isdir, sticky, disallow, fullpath) tmp_file_sec(dir_val, isdir, sticky, disallow, fullpath)
+#define STAT_CHKPT_FILE(file_name, buf) stat(file_name, buf)
```
Comment (subhasisb) [inline src/include/portability.h:58]: do have this, instead implement stat in libwin

File: src/resmom/mom_main.c [hunk 6]
```diff
@@ -4755,17 +4275,13 @@ read_config(char *file)
 			return 0;	/* ok for "config" not to be there  */
 	}
 #if !defined(DEBUG) && !defined(NO_SECURITY_CHECK)
-#ifdef	WIN32
-	if (chk_file_sec(file, 0, 0, WRITES_MASK^FILE_WRITE_EA, 0)) {
+	perm = get_permission("write");
+	if (chk_file_sec(file, 0, 0, perm, FULLPATH)) {
```
Comment (subhasisb) [inline src/resmom/mom_main.c:4279]: Why does this have to be different between the two ?

**Review by subhasisb — CHANGES_REQUESTED** [review]

File: src/resmom/mom_main.c (hunk 6 above)
Comment (SudeshnaMoulik) [reply src/resmom/mom_main.c:4279]: As per the current design, only for *nix platform we check the full path. And if we see the whole code-base for chk_file_sec(), depending on the file/path/code flow this also varies. 
So to make it consistent with current behavior used the macro

**Review by subhasisb — APPROVED** [review]

**Comment by SudeshnaMoulik** [comment]
Putting this PR in WIP state to resolve conflicts that I got after re-basing with master.

**Review by npadole20 — APPROVED** [review]
Looks good to me.

**Comment by SudeshnaMoulik** [comment]
Latest logs:
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|3800|3550|17|1|1|3|3528|

**Review by Shrini-h — APPROVED** [review]
LGTM, Thank you for making the changes

**Review by subhasisb — APPROVED** [review]

**Pull request merged**